zerver/webhooks/groove/tests.py

```diff
@@ -0,0 +1,124 @@
+    # This test simulates the condition when
+    # a ticket is assigned to no one (it
+    # would be ignored)
+    def test_groove_ticket_assigned4(self) -> None:
+        self.subscribe(self.test_user, self.STREAM_NAME)
+        result = self.client_post(self.url, self.get_body('ticket_assigned4'),
+                                  content_type="application/x-www-form-urlencoded",
+                                  X_GROOVE_EVENT='ticket_assigned')
+        self.assert_json_success(result)
+
+    # This test simulates the notification when
+    # an agent replied to a ticket
+    def test_groove_agent_replied(self) -> None:
+        expected_subject = u"notifications"
+        expected_message = u"agent@example.com has just replied to a ticket\n" \
+                           u"```quote\n" \
+                           u"**[Ticket #776](https://ghostfox.groovehq.com/groove_client/tickets/68667295)**\n" \
+                           u"Hello , This is a reply from an agent to a ticket\n" \
+                           u"```"
+
+        self.send_and_test_stream_message('agent_replied', expected_subject, expected_message,
+                                          content_type="application/x-www-form-urlencoded",
+                                          X_GROOVE_EVENT="agent_replied")
+
+    # This test simulates the notification when
+    # a customer replied to a ticket
+    def test_groove_customer_replied(self) -> None:
+        expected_subject = u"notifications"
+        expected_message = u"rambo@example.com has just replied to a ticket\n" \
+                           u"```quote\n" \
+                           u"**[Ticket #440](https://ghostfox.groovehq.com/groove_client/tickets/68666538)**\n" \
+                           u"Hello agent, thanks for getting back. This is how a reply from customer looks like.\n" \
+                           u"```"
+
+        self.send_and_test_stream_message('customer_replied', expected_subject, expected_message,
+                                          content_type="application/x-www-form-urlencoded",
+                                          X_GROOVE_EVENT="customer_replied")
+
+    # This test simulates the notification when
+    # an agent left a note
+    def test_groove_note_added(self) -> None:
+        expected_subject = u"notifications"
+        expected_message = u"anotheragent@example.com has left a note\n" \
+                           u"```quote\n" \
+                           u"**[Ticket #776](https://ghostfox.groovehq.com/groove_client/tickets/68667295)**\n" \
+                           u"This is a note added to  a ticket\n" \
+                           u"```"
+
+        self.send_and_test_stream_message('note_added', expected_subject, expected_message,
+                                          content_type="application/x-ww-form-urlencoded",
+                                          X_GROOVE_EVENT="note_added")
+
+    # This test is for other events (it would
+    # be ignored)
```

Comment:
Why do we ignore other types of events? We already have a fixture for the `ticket_state_changed` event.

Reply:
There are a lot of other events in Groove (there are fifteen in total). However, when Groove is integrated with Slack, we only received eight types of notifications, six of which are identifiable from which event it came from, and the last two are not (I experimented it myself and also consulted to the research note [here](https://github.com/zulip/zulip-gci-submissions/blob/master/webhook-integrations/groove/notes.md)). `ticket_state_changed` is not one of the identified events, and when triggered it does not make a notification message in Slack. Since I'm basing this on the integration in Slack, I treated it as the other types of events and just ignored it.

However, if you wish to include it too as part of the notifications, let me know. There are no messages that can be based on though, so I would write the format of the messages on my own.